src/connector/src/sink/formatter/mod.rs

```diff
@@ -68,13 +69,22 @@ macro_rules! tri {
 }
 
 pub enum SinkFormatterImpl {
+    // append-only
     AppendOnlyJson(AppendOnlyFormatter<JsonEncoder, JsonEncoder>),
+    AppendOnlyTextJson(AppendOnlyFormatter<TextEncoder, JsonEncoder>),
     AppendOnlyProto(AppendOnlyFormatter<JsonEncoder, ProtoEncoder>),
+    AppendOnlyTextProto(AppendOnlyFormatter<TextEncoder, ProtoEncoder>),
+    AppendOnlyTemplate(AppendOnlyFormatter<TemplateEncoder, TemplateEncoder>),
+    AppendOnlyTextTemplate(AppendOnlyFormatter<TextEncoder, TemplateEncoder>),
+    // upsert
     UpsertJson(UpsertFormatter<JsonEncoder, JsonEncoder>),
+    UpsertTextJson(UpsertFormatter<TextEncoder, JsonEncoder>),
     UpsertAvro(UpsertFormatter<AvroEncoder, AvroEncoder>),
-    DebeziumJson(DebeziumJsonFormatter),
-    AppendOnlyTemplate(AppendOnlyFormatter<TemplateEncoder, TemplateEncoder>),
+    UpsertTextAvro(UpsertFormatter<TextEncoder, AvroEncoder>),
     UpsertTemplate(UpsertFormatter<TemplateEncoder, TemplateEncoder>),
+    UpsertTextTemplate(UpsertFormatter<TextEncoder, TemplateEncoder>),
```

Comment:
So here we expand the cartesian product of 

```
{Append-only, Upsert} x {Text (key) or not} x {Json, Proto, Template}
```

Is it possible to separate them?

Reply:
Possibly. The caller of formatter does not need to know whether it is `AvroEncoder` or `JsonEncoder`, but it still need to know whether the output of an encoder `SerTo` a (utf8) `String` or not (`Vec<u8>`). So it may be possible to reduce into `{Append-only, Upsert} x {String, Vec<u8>} x {String, Vec<u8>}` (or possibly `{Append-only, Upsert} x {String, Vec<u8>} x {Vec<u8>}` if all MQs allow payload to be non-utf8 bytes). But the refactor wont fit into this PR. I will give it a try after this.